037-auto-rebalance-cluster-scaling.md

```diff
@@ -0,0 +1,177 @@
+
+## Current situation
+
+Currently, there is no automatic rebalancing when the user scales the cluster up or down.
+It means that:
+
+* after scaling up the cluster, the user has to create a `KafkaRebalance` custom resource with the `spec.mode: add-brokers` and the `spec.brokers` with the added brokers, in order to rebalance the cluster and moving some replicas to the newly added brokers.
+* before scaling down the cluster, the user has to create a `KafkaRebalance` custom resource with the `spec.mode: remove-brokers` and the `spec.brokers` with the brokers to remove, in order to rebalance the cluster and moving replicas out of the brokers to be removed.
+
+## Motivation
+
+The rebalancing operation during scaling up and down is a manual two step process.
+
+* scaling up the cluster, then running a rebalance.
+* running a rebalance, then scaling down the cluster.
+
+In order to simplify, it would be great for the user to deal with just scaling the cluster by changing the `spec.kafka.replicas` field and then having the operator taking care to run a rebalancing operation automatically if requested.
+
+## Proposal
+
+We would need to extend the `Kafka` custom resource to allow the user specifying that he wants auto-rebalancing on scaling.
+
+A new `spec.kafka.autoRebalance` field could be added for this purpose.
+Because we are talking about auto-rebalancing, the corresponding field could also be not limited to the rebalancing on scaling but a more general “continuous” auto-rebalancing to enable on the cluster. 
+To specify when running the auto-rebalancing, a `spec.kafka.autoRebalance.mode` field could be used by listing if running auto-rebalancing only on scaling, during the normal execution or both.
+Having “continuous” auto-rebalancing is out of scope for this proposal, the field is added for allowing a future implementation.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    replicas: 3
+    # ...
+    autoRebalance:
+      mode: [running, scaling] # or just one of the two
+```
+
+Finally, the user should be able to define what are the [goals and the other rebalancing related configurations](https://strimzi.io/docs/operators/latest/configuring.html#type-KafkaRebalance-reference) which are normally described via a `KafkaRebalance` custom resource.
+In any case, for rebalancing the cluster, an actual `KafkaRebalance` custom resource has to be automatically created by the cluster operator.
+
+### New `KafkaRebalanceTemplate` custom resource
+
+A new `KafkaRebalanceTemplate` custom resource could be a solution allowing the user to define the rebalancing configuration he wants but leaving to the operator the creation of the corresponding `KafkaRebalance` custom resource.
+This new CRD should be quite similar to the `KafkaRebalance`.
+The cluster operator has to copy over the configuration from the template to the actual rebalancing resource.
```

Comment:
My point is that the user wants anyway to define specific configuration for rebalance operations which happen on scale up/down. This is where the `KafkaRebalanceTemplate` comes into the picture and is used to allow the operator to generate the corresponding `KafkaRebalance`. If not leveraging the current custom resource how the user would be able to configure them? Again, back to one of the rejected alternatives putting everything in the `Kafka` CR? What could be a "different input"?
Anyway the auto generated `KafkaRebalance` is named with something like `auto-rebalance` suffix and of course documentation will help as well.

Reply:
I agree that the user will need a mechanism to configure the rebalance operations they want to use during a rebalance, so the purpose of the `KafkaRebalanceTemplate` is clear to me. My concern is about the `KafkaRebalance` CRs that are then created as a result. Will it be clear to the user how to interact with the `KafkaRebalance` CRs created by them vs the ones created by the Strimzi operator. For example the `KafkaRebalance` resources that are suffixed with `auto-rebalance` shouldn't be edited by them, the user shouldn't change the config there or add annotations, they should change the config in the `KafkaRebalanceTemplate` instead and create their own `KafkaRebalance` if the automated one fails. This flow to me feels like it is open to confusion from users.

Having said that I haven't been able to think of a flow that would be as simple to implement from a Strimzi perspective. Without the intermediate `KafkaRebalance` you would need to track the status of the rebalancing in a different way.

Reply:
One thing to keep in mind is GitOps tooling. Ideally we want to avoid it deleting these automatically created KafkaRebalance resources :-o.

Reply:
Creating the KafkaRebalance resource manually can be done today. That does not need any changes or any proposal. The point of this proposal is to do it automatically, without any manual action from the user.

Reply:
If I get this right, the user manually creates the template, so why not directly create the KafkaRebalance? Then, when replicas change, the operator will enrich it with whatever is needed for the specific mode.

Reply:
Yeah, but the idea is that you for example create you template once when creating your Kafka cluster. And this is used few months later when you scale it up or down. Not that you create the template manually few seconds before the scaling.

Reply:
Yeah, that's what I meant. The operator should be smart enough to filter them out when it's not a scale up/down situation.

Reply:
I think that would be quite confusing for the user and for managing the Rebalance resources when it mixes some used as templates and some used as actual rebalances.

Reply:
> If I get this right, the user manually creates the KafkaRebalanceTemplate, so why not directly create the KafkaRebalance? Then, when replicas change, the operator will enrich it with whatever is needed for the specific mode.

So `the operator will enrich it` is one of the things that an operator should avoid to do imho. In general, an operator touches/changes only annotations on a resource it's handling (because of the management itself) but it doesn't add/update anything about the `spec` section which is defined by the user. The operator just updates the `status` one.
We could have an already filled `KafkaRebalance` with mode as well but we should think a way to allow the operator to skip it on reconcile if it's not the time to use it (because no scaling up/down). Maybe a `trigger` field could help but anyway it will come to the point by Jakub which I agree (users confused by mixing KafkaRebalance resources).
The other thing is that you can share a template for scaling up and down. Using `KafkaRebalance` directly would mean creating two identical custom resources with a different mode (which is in the end what the operator does for you behind the scenes).

> I think that would be quite confusing for the user and for managing the Rebalance resources when it mixes some used as templates and some used as actual rebalances.

I agree